File: my-folder/0417-pacific-atlantic-water-flow/solution.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        rows = len(heights)
        cols = len(heights[0])
        pacific = set()
        atlantic = set()

        def dfs(r, c, visit, prevHeight):
            if (r, c) in visit or r < 0 or c < 0 or r >= rows or c >= cols or heights[r][c] <  prevHeight:
                return

            visit.add((r, c))
            
            currHeight = heights[r][c]
            dfs(r + 1, c, visit, currHeight)
            dfs(r - 1, c, visit, currHeight)
            dfs(r, c + 1, visit, currHeight)
            dfs(r, c - 1, visit, currHeight)

        for col in range(cols):
            dfs(0, col, pacific, heights[0][col])
            dfs(rows - 1, col, atlantic, heights[rows - 1][col])
        
        for row in range(rows):
            dfs(row, 0, pacific, heights[row][0])
            dfs(row, cols - 1, atlantic, heights[row][cols - 1])
        res = []

        for r in range(rows):
            for c in range(cols):
                if (r,c) in pacific and (r,c) in atlantic:
                    res.append([r,c])
        return res

        

File: my-folder/0417-pacific-atlantic-water-flow/solution.java
class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        int rows = heights.length;
        int cols = heights[0].length;
        boolean[][] pacific = new boolean[rows][cols];
        boolean[][] atlantic = new boolean[rows][cols];
        
        // Run DFS from all ocean-bordering cells
        for (int c = 0; c < cols; c++) {
            dfs(0, c, pacific, heights[0][c], heights);
            dfs(rows - 1, c, atlantic, heights[rows - 1][c], heights);
        }
        for (int r = 0; r < rows; r++) {
            dfs(r, 0, pacific, heights[r][0], heights);
            dfs(r, cols - 1, atlantic, heights[r][cols - 1], heights);
        }

        // Collect results
        List<List<Integer>> res = new ArrayList<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (pacific[r][c] && atlantic[r][c]) {
                    res.add(Arrays.asList(r, c));
                }
            }
        }
        return res;
    }

    private void dfs(int row, int col, boolean[][] visit, int prevHeight, int[][] heights) {
        int rows = heights.length;
        int cols = heights[0].length;

        // Boundary + visited + invalid height check
        if (row < 0 || row >= rows || col < 0 || col >= cols || visit[row][col] || heights[row][col] < prevHeight) {
            return;
        }

        visit[row][col] = true;
        int currHeight = heights[row][col];

        dfs(row + 1, col, visit, currHeight, heights);
        dfs(row - 1, col, visit, currHeight, heights);
        dfs(row, col + 1, visit, currHeight, heights);
        dfs(row, col - 1, visit, currHeight, heights);
    }
}

